[ADP-322] Add metadata encryption to HTTP API.

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->


- [x] Adding  noninvasive api changes to `constructTransaction`
- [x] Applying changes to Api.Server 
- [x] Extend `ErrConstructTx` 
- [x] Implementing `toMetadataEncryption` according to https://github.com/cardano-foundation/CIPs/tree/master/CIP-0083
- [x] Prepare goldens in line to openSSL and CIP83 examples
- [x] Show error emitting works
- [x] Add salting to key gen and encryption
- [x] Add integration test showing error 
- [x] Add integration test showing encryption

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number
adp-322

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
update swagger

paweljakubas committed:
introduce EncryptMetadataMethod

paweljakubas committed:
add ErrConstructTxIncorrectRawMetadata

paweljakubas committed:
more logic in toMetadataEncrypted

paweljakubas committed:
even more logic in toMetadataEncrypted

paweljakubas committed:
add golden 1

paweljakubas committed:
add golden 2

paweljakubas committed:
the rest of golden - salt used 2

paweljakubas committed:
be in line with 674 label

paweljakubas committed:
error testing

paweljakubas committed:
typos

paweljakubas committed:
integration tests 1

paweljakubas committed:
hlint

paweljakubas committed:
add genSalt and use it in constructTransaction

paweljakubas committed:
integration tests 2

paweljakubas committed:
getSaltFromEncrypted

paweljakubas committed:
metadata integration test with long msg

paweljakubas committed:
better code reuse in integration metadata encryption tests

jonathanknowles committed:
Miscellaneous formatting fixes.

This commit makes the following fixes to formatting:

  - Replace hanging vertical alignment with newline characters and
    fixed-width indentation.

  - Wrap lines to fit within the 80-column line length limit.

  - Move closing parentheses to the same column as opening parentheses for
    multi-line parenthetical expressions.

jonathanknowles committed:
Extract out common constant `saltLengthBytes`.

This helps us to avoid repeating the magic number `8` in multiple
places.

jonathanknowles committed:
Use `stripPrefix` to simplify `getSaltFromEncrypted`.

jonathanknowles committed:
Use pattern guards in `getSaltEncrypted`.

The use of a `do` block is unnecessary.

jonathanknowles committed:
Move salt-related constants and functions together.

jonathanknowles committed:
Use `stripPrefix` to simplify `decrypt`.

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -163,3 +164,13 @@ decrypt mode key iv msg = do
+getSaltFromEncrypted msg = do
+    when (BS.length msg < 32) Nothing
+    let (prefix,rest) = BS.splitAt 8 msg
+    let saltDetected = prefix == saltPrefix
```
The combination of `splitAt` with `== saltPrefix` can be replaced with `stripPrefix`.

See: https://github.com/cardano-foundation/cardano-wallet/pull/4581

jonathanknowles committed:
Rename type to `ApiEncryptMetadataMethod`.

This follows the convention where API types are prefixed with `Api`.

jonathanknowles committed:
Add `Arbitrary` instance for `ApiEncryptMetadataMethod`.

jonathanknowles committed:
Simplify `Arbitrary` instance for `ApiEncryptMetadata`.

jonathanknowles committed:
Add JSON golden test for `ApiEncryptMetadata`.

jonathanknowles committed:
Change name of field from `enc` to `method`.

jonathanknowles committed:
Rename function `method` to `stdMethod`.

jonathanknowles committed:
Rename `ApiEncryptMetadataMethod` constructor to `Basic`.

jonathanknowles committed:
Derive `{To,From}JSON` instances for `ApiEncryptMetadataMethod`.

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1237,8 +1238,23 @@ data ApiMultiDelegationAction
+
+data ApiEncryptMetadata = ApiEncryptMetadata
+    { passphrase :: ApiT (Passphrase "lenient")
+    , enc :: Maybe EncryptMetadataMethod
```
Hi @paweljakubas 

For what reason do we allow this field to be `Nothing`?

If it is set to `Nothing`, then what do we expect the behaviour to be?

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1237,8 +1238,23 @@ data ApiMultiDelegationAction
+        if txt == "basic" then
+            pure AES256CBC
+        else
+            fail "'basic' is expected."
```
Suggestion: derive this instance automatically.

See: https://github.com/cardano-foundation/cardano-wallet/pull/4588

jonathanknowles commented on lib/api/src/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -3121,6 +3153,96 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
+    merge Nothing (Just val) = Just val
+    merge (Just val) Nothing = Just val
+    merge Nothing Nothing = Nothing
+    merge (Just _) (Just _) = error "only one 'msg' field expected"
```
@paweljakubas is it possible to trigger this run-time error from the API?

I've created an experimental test to test this hypothesis here:
https://github.com/cardano-foundation/cardano-wallet/pull/4601

jonathanknowles replied:
Conclusion: I think it's impossible to trigger this call to `error` from the API.

This is because the JSON serialisation for `TxMetaMap` ensures that there can never be more than one map entry with any given key.

(It's unfortunate that this restriction is not reflected in the choice of type for `TxMetaMap`, which is just a list of pairs.)

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -1788,17 +1788,29 @@ x-transactionMetadata: &transactionMetadata
+    Cipher algorithm uses 8-byte salt, PKCS#7 padding as specified in
+    https://datatracker.ietf.org/doc/html/rfc5652#section-6.3 is applied.
+    Only metadata value under `msg` field is encrypted. If `msg` field is missing error
+    will be emitted.
```
As a user who is new to this feature, I'm finding this `description` section a little hard to follow!

For example, it refers to a field called `msg`. However, this object only has two fields: `enc` and `passphrase`.

From reading through the tests (`Api.TypesSpec`), I can see that `msg` is supposed to be the key in a `TxMetaMap`.

Is there any way that we can give a more complete description of how to use this feature? (Ideally, so that the user doesn't have to go hunting through CIP 83.)

jonathanknowles committed:
Fix up imports of `ApiEra`.

jonathanknowles committed:
Remove unused `getMetadata` functions.

jonathanknowles committed:
Fix formatting of comment for `toMetadataEncrypted`.

jonathanknowles committed:
Add type signatures to inner definitions of `toMetadataEncrypted`.

jonathanknowles committed:
Use unqualified imports for `TxMeta` and `TxMetadataValue`.

jonathanknowles committed:
Shorten definition of `updateTxMetadata`.

jonathanknowles committed:
Use `LambdaCase` in `encryptingMsg`.

jonathanknowles committed:
Simplify definition of `encryptPairIfQualifies`.

jonathanknowles committed:
Use `LambdaCase` in `inspectMetaPair`.

jonathanknowles committed:
Simplify definition of `getMsgValue`.

jonathanknowles committed:
Create definition for `flip toTextChunks []`.

jonathanknowles committed:
Use `toTextChunks` to refer to outer function.

jonathanknowles committed:
Extract out hard-coded `chunkSize` constant within `toTextChunks`.

jonathanknowles committed:
Replace `toTextChunks` with `Text.chunksOf`.

These functions are identical.

jonathanknowles committed:
Use `where` instead of `let` and `in` for `encryptPairIfQualifies`.

The result is more concise, saving both horizontal and vertical space.

jonathanknowles committed:
Make `getMsgValue` an inner function of `inspectMetaPair`.

jonathanknowles committed:
Make `merge` an inner function of `inspectMetaPair`.

jonathanknowles committed:
Simplify function `inspectMetaPair`.

jonathanknowles committed:
Further simplify function `getMsgValue`.

jonathanknowles committed:
Extract out constant `cip20MetadataKey`.

jonathanknowles committed:
Restrict return type of `findMsgValue`.

jonathanknowles committed:
Use pattern guards in function `findMsgValue`.

jonathanknowles committed:
Further restrict return type of `findMsgValue`.

There can only be a single (k, v) pair, since each unique key maps to at
most one value.

jonathanknowles committed:
Defer creation of (redundant) singleton list.

jonathanknowles committed:
Further defer creation of (redundant) singleton list.

jonathanknowles committed:
Eliminate creation of redundant singleton list.

jonathanknowles committed:
Stop unnecessarily passing `cip20MetadataKey` around.

The key returned by `findMsgValue` was always equal to the value of the
`cip20MetadataKey` constant. Other functions that need to refer to this
constant can do so directly.

jonathanknowles committed:
Remove unnecessary alias of "msg".

jonathanknowles committed:
Inline definition of `encMethodEntry`.

jonathanknowles committed:
Extract out definition `encryptedChunks`.

jonathanknowles committed:
Inline definition of `encrypted`.

jonathanknowles committed:
Replace `mapBoth` with `bimap`.

The `bimap` function is already widely used within this codebase, and is
exported from `base`, whereas `mapBoth` requires the use of an
additional library.

jonathanknowles committed:
Inline definition of `toBase64Text`.

jonathanknowles committed:
Rename function to `parseMessage`.

jonathanknowles committed:
Rename function to `getValue`.

jonathanknowles committed:
Rename function to `validKeyAndMessage`.

jonathanknowles committed:
Rename function to `extractMessage`.

jonathanknowles committed:
Rename function to `encryptMessage`.

jonathanknowles committed:
Simplify definition of function `encryptMessage`.

jonathanknowles committed:
Make `encryptPairIfQualifies` an inner function.

Of `encryptMessage`.

jonathanknowles committed:
Extract out function `encryptValue`.

jonathanknowles committed:
Flatten `where` hierarchy for function `encryptMessage`.

jonathanknowles committed:
Add type signatures to inner functions of `toMetadataEncrypted`.

jonathanknowles committed:
Simplify definition of `toMetadataEncrypted`.

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4732,6 +4744,17 @@ x-errWithdrawalNotPossibleWithoutVote: &errWithdrawalNotPossibleWithoutVote
+  properties:
+    message:
+      type: string
+      description: The metadata has either wrong structure or cannot be encrypted.
```
```suggestion
      description: |
          The supplied metadata object is not compatible with the format
          specified by CIP-83 (https://cips.cardano.org/cip/CIP-83).
```

paweljakubas replied:
done

jonathanknowles reviewed (approved):
Hi @paweljakubas 

Many thanks for creating this PR and for your patience with waiting for reviews.

Given this PR adds just the **_encryption_** functionality, with **_decryption_** planned for a later PR, I think this PR is okay to merge for now.

However, I do have some proposals that I think should be addressed in subsequent PRs:

1. Currently, the "business logic" for metadata encryption is located within the `Cardano.Wallet.Api.Http.Shelley.Server` module. I would propose to extract this logic out into a pure function that (a). doesn't rely on HTTP API types and (b). lives in a more appropriate location, preferably outside of the `Server` module, so that the `Server` module only needs to handle concerns that directly relate to the HTTP API. Perhaps this new location could be something like `Cardano.Write.Tx.Metadata.Encryption`?

2. I would propose to replace all instances of the magic value `674` with references to the `cip20MetadataKey` constant (already used by `toMetadataEncrypted`), and similarly, to re-locate this constant to the proposed `Cardano.Write.Tx.Metadata.Encryption` module.

3. When we have a pure function that performs **_decryption_**, I would propose to:
    1. add a new test module `Cardano.Write.Tx.Metadata.EncryptionSpec`.
    1. add a QuickCheck property test for the encryption-decryption round-trip property: `decrypt . encrypt == id`.
    1. add a selection of golden example-based unit tests (i.e., tests that are very quick to run that don't require us to fire up the integration test suite).

4. I still find the description of `x-encryptMetadata` in the swagger file a little hard to follow. Perhaps it would be possible to find someone who is completely new to this feature, and work with them to revise this description in a way that is easier for new users to understand? (See earlier review comment: https://github.com/cardano-foundation/cardano-wallet/pull/4555/files#r1599427620)

Again, many thanks again for creating this PR!

Jonathan

paweljakubas committed:
swagger improve
